[Fix][Printer] Fix PrintFinal dispatch for Relax dedicated Expr

MasterJH5574 opened the pull request:
PR #306 introduces dedicated Relax Exprs, making Relax sub-class Expr independent of Relay Expr. This PR is a fix on the text printer accordingly.

Prior to this PR, TextPrinter::PrintFinal only dispatch VarNode, FunctionNode and ShapeExprNode to Relax printer, leaving all other kinds of Exprs to Relay text printer. This is fine at the time before PR #306, when we were reusing Exprs from Relay. But since #306 this becomes an issue, because Relax Exprs are not properly dispatched to Relax printer. As a result, when we print a CallNode or TupleNode on Python side, or structural-equality check finds mismatch on Tuples, the PrettyPrint as well as PrintFinal cannot dispatch the Expr, and will result in error like
```
  Check failed: (can_dispatch(n)) is false: NodeFunctor calls un-registered function on type relax.expr.Tuple
```
which is not supposed and is also the least informative.

So this PR fixes the issue by adding our Relax Exprs to the dispatch rule. There are corresponding unit tests to make sure the printer issue is not likely to happen again.

---

cc @Hzfengsy @YuchenJin @tqchen

MasterJH5574 commented:
Waiting for #349 to merge first and update the ConstantNode unit test.

tqchen commented:
one note  perhaps we need to move PrintFatal to dispatch to script printing later. (cc @junrushao)

MasterJH5574 commented:
This one is now ready for review.

YuchenJin commented on tests/python/relax/test_printer.py:
```diff
@@ -449,5 +450,130 @@ def local_func_3(
+
+    const0 = relax.const(1.0, "float32")
+    assert isinstance(const0, relax.Constant)
+    assert const0.__str__() == 'R.const(1, "float32")'
```
testing scalar is good, can we also add a slightly more complicated tensor case?

MasterJH5574 replied:
I tried high-dimensional tensor but the string comparison always fails. Give me some time to investigate.

YuchenJin commented on tests/python/relax/test_printer.py:
```diff
@@ -449,5 +450,130 @@ def local_func_3(
+
+    block = foo0.body.blocks[0]
+    assert isinstance(block, relax.BindingBlock)
+    block.__str__()
```
Do we want to compare with expected string?

MasterJH5574 replied:
Yes we can. The expected string is two-line long so I will probably need raw string for help.

MasterJH5574 commented:
Thanks @YuchenJin! Just addressed your comments with one more case of constants, and string comparison for BindingBlock and SeqExpr. Please take another look :-)

YuchenJin commented:
Thanks @MasterJH5574 for this patching the printing, it has bothered many of us for a while.
